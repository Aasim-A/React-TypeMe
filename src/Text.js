import React, { useState, useEffect } from 'react';

const TYPING_SPEED = 10; // characters per second
const DELETE_SPEED = 20; // cps
const CHAR_INTERVAL = 1000 / TYPING_SPEED; // ms
const DEL_INTERVAL = 1000 / DELETE_SPEED;
const DELETE_DELAY = 2000; // ms

const isServer = () => {
  return !(typeof window != 'undefined' && window.document);
};

const Text = ({
  endDelay,
  className,
  children,
  lineBreak,
  keepCursor,
  onAnimationEnd,
  startAnimation,
  cursorCharacter,
  deleteCharacters
}) => {
  const [string, setString] = useState('');
  const [delayed, setDelayed] = useState(false);
  const [deleting, setDeleting] = useState(false);
  const [animationEnded, setAnimationEnded] = useState(false);
  const childrenLength = children.length;
  let elapsed = 0;

  useEffect(() => {
    const updateString = time => {
      if (elapsed === 0) {
        elapsed = time;
      }
      let curLength = string.length;
      if (!deleting && curLength < childrenLength) {
        if (time >= elapsed + CHAR_INTERVAL) {
          setString(children.substring(0, curLength + 1));
          elapsed = time;
        } else {
          window.requestAnimationFrame(updateString);
        }
      } else {
        if (deleteCharacters > 0) {
          if (!deleting) {
            setDelayed(true);
            window.setTimeout(() => {
              setDeleting(true);
              setDelayed(false);
              setString(children.substring(0, curLength - 1));
            }, DELETE_DELAY);
          } else {
            if (curLength === childrenLength - deleteCharacters) {
              if (endDelay > 0) {
                setDelayed(true);
                window.setTimeout(() => {
                  setDelayed(false);
                  setAnimationEnded(true);
                  onAnimationEnd();
                }, endDelay);
              } else {
                setAnimationEnded(true);
                onAnimationEnd();
              }
            } else {
              if (time >= elapsed + DEL_INTERVAL) {
                setString(children.substring(0, curLength - 1));
              } else {
                window.requestAnimationFrame(updateString);
              }
            }
          }
        } else {
          // animation end
          if (endDelay > 0) {
            setDelayed(true);
            window.setTimeout(() => {
              setDelayed(false);
              setAnimationEnded(true);
              onAnimationEnd();
            }, endDelay);
          } else {
            setAnimationEnded(true);
            onAnimationEnd();
          }
        }
      }
    };
    if (startAnimation && !isServer()) {
      window.requestAnimationFrame(updateString);
    }
  }, [string, startAnimation]);

  let cn = `${className}${className.length > 0 ? ' ta-cursor' : 'ta-cursor'}`;
  cn = `${cn}${animationEnded || delayed ? ' ta-blink' : ''}`;
  if (!startAnimation && !animationEnded) {
    cn = `${cn} ta-hide`;
  } else {
    cn = `${cn}${animationEnded && !keepCursor ? ' ta-hide' : ''}`;
  }
  return (
    <span className="ta-anim">
      <span className={`${className}`}>{isServer() ? children : string}</span>
      {!isServer() && <span className={cn}>{cursorCharacter}</span>}
      {lineBreak ? <br /> : null}
    </span>
  );
};

Text.defaultProps = {
  className: '',
  eraseLength: 0,
  lineBreak: false,
  keepCursor: true,
  endDelay: false,
  startAnimation: true,
  onAnimationEnd: () => {
    console.log('onAnimationEnd => 0');
  }
};

export default Text;