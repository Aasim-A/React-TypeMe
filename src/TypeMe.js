import React, { useState, useEffect } from 'react';

const FORWARD = 1;
const BACKSPACE = -1;
const LINEBREAK = 2;
const PAUSE = 3;
const END = 0;

const TypeMe = ({
  loop,
  strings,
  children,
  className,
  hideCursor,
  typingSpeed,
  deleteSpeed,
  backspaceDelay,
  startAnimation,
  onAnimationEnd,
  cursorCharacter,
  cursorBlinkSpeed
}) => {
  const [instanceId] = useState(
    () =>
      Math.random()
        .toString(36)
        .substring(2, 5) +
      Math.random()
        .toString(36)
        .substring(2, 5)
  );
  const [deleteChar, setDeleteChar] = useState(0);
  const [itemIndex, setItemIndex] = useState(0);
  const [charIndex, setCharIndex] = useState(0);
  const [animationPaused, setAnimationPaused] = useState(false);
  const [animationEnded, setAnimationEnded] = useState(false);
  const [typedString, setTypedString] = useState('');
  const [newTypedString, setNewTypedString] = useState('');
  const typingInterval = (1000 * 60) / (typingSpeed * 5); // ms
  const deleteInterval = (1000 * 60) / (deleteSpeed * 5); // ms

  let nextItem;
  let elapsed = 0;

  const getNextItem = items => {
    let item;
    if (itemIndex >= items.length) {
      return {
        direction: END // animation ends
      };
    }
    item = items[itemIndex];
    if (typeof item === 'string') {
      return {
        direction: FORWARD,
        string: item
      };
    }
    switch (item.type.displayName) {
      case 'LineBreak':
        return {
          direction: LINEBREAK
        };

      case 'Delete':
        let delay = false;
        let newDeleteChar = 0;
        if (deleteChar === 0) {
          if (item.props.characters === 0) {
            newDeleteChar = newTypedString.length;
          } else {
            newDeleteChar = item.props.characters;
          }
          setDeleteChar(newDeleteChar);
          delay = true;
        } else {
          newDeleteChar = deleteChar - 1;
          setDeleteChar(newDeleteChar);
        }
        return {
          delay,
          direction: BACKSPACE,
          chars: newDeleteChar
        };

      case 'Delay':
        return {
          direction: PAUSE,
          ms: item.props.ms
        };

      default:
        throw 'Error: Invalid item passed in `strings` props or as children.';
    }
  };

  const updateTypedString = (interval, nts) => {
    return () => {
      let time = performance.now();
      if (elapsed === 0) {
        elapsed = time;
      }
      if (time >= elapsed + interval) {
        elapsed = time;
        const split = nts.split('•');
        setTypedString(
          split.map((str, index) => {
            return (
              <span key={`${instanceId}-${index}`}>
                {str}
                {split.length - index > 1 ? <br /> : null}
              </span>
            );
          })
        );
      } else {
        window.setTimeout(updateTypedString(interval, nts), interval);
      }
    };
  };

  useEffect(() => {
    if (window && !window._TYPEME) {
      let styleSheet = document.styleSheets[0];
      if (!styleSheet) {
        const style = document.createElement('style');
        document.head.append(style);
        styleSheet = style.sheet;
      }
      let animName = 'tm-blink';
      let animStyle = `{0%{opacity:1;}49%{opacity:1;}50%{opacity:0;}100%{opacity:0;}}`;
      let keyframes = [
        `@keyframes ${animName}${animStyle}`,
        `@-webkit-keyframes ${animName}${animStyle}`,
        `.tm-cursor{display:inline-block;transform:scale(1.2);font:inherit;position:relative;font-style:normal !important;}`,
        `.tm-blink{animation:${animName} ${cursorBlinkSpeed}ms infinite !important;}`,
      ];
      keyframes.forEach(style => {
        styleSheet.insertRule(style, styleSheet.cssRules.length);
      });
      window._TYPEME = true;
    }
  }, []);

  useEffect(() => {
    if (startAnimation) {
      let items = [];
      if (strings && Array.isArray(strings)) {
        items = strings;
      }
      if (strings && typeof strings === 'string') {
        items = [strings];
      }
      if (children && typeof children === 'string') {
        items = [children];
      }
      if (children && Array.isArray(children)) {
        items = children;
      }
      nextItem = getNextItem(items);
      let { direction } = nextItem;
      if (direction === END) {
        onAnimationEnd();
        if (loop) {
          setCharIndex(0);
          setItemIndex(0);
          setTypedString('');
          setNewTypedString('');
        } else {
          setAnimationEnded(true);
        }
      } else {
        if (direction === FORWARD) {
          // type next character
          let nts = `${newTypedString}${nextItem.string[charIndex]}`;
          setNewTypedString(nts);
          window.setTimeout(
            updateTypedString(typingInterval, nts),
            typingInterval
          );
          if (charIndex >= nextItem.string.length - 1) {
            setCharIndex(0);
            setItemIndex(prevIndex => prevIndex + 1);
          } else {
            setCharIndex(prevIndex => prevIndex + 1);
          }
        } else if (direction === LINEBREAK) {
          // break line
          let nts = `${newTypedString}•`;
          setNewTypedString(nts);
          window.setTimeout(
            updateTypedString(typingInterval, nts),
            typingInterval
          );
          setItemIndex(prevIndex => prevIndex + 1);
          setCharIndex(0);
        } else if (direction === BACKSPACE) {
          // delete previous character
          let nts = `${newTypedString.substring(0, newTypedString.length - 1)}`;
          setNewTypedString(nts);
          if (nextItem.chars === 1) {
            setItemIndex(prevIndex => prevIndex + 1);
            setCharIndex(0);
            setDeleteChar(0);
          }
          if (nextItem.delay) {
            window.setTimeout(() => {
              window.setTimeout(
                updateTypedString(deleteInterval, nts),
                deleteInterval
              );
            }, backspaceDelay);
          } else {
            window.setTimeout(
              updateTypedString(deleteInterval, nts),
              deleteInterval
            );
          }
        } else if (direction === PAUSE) {
          // pause animation
          setItemIndex(prevIndex => prevIndex + 1);
          setCharIndex(0);
          window.setTimeout(() => {
            setAnimationPaused(false);
            window.setTimeout(
              updateTypedString(typingInterval, newTypedString),
              typingInterval
            );
          }, nextItem.ms);
          setAnimationPaused(true);
        }
      }
    }
  }, [startAnimation, typedString]);

  let containerCn = 'tm';
  let cursorCn = 'tm-cursor';
  if (className) {
    containerCn = `${containerCn} ${className}`;
  }
  if (!startAnimation || animationEnded || animationPaused) {
    cursorCn = `${cursorCn} tm-blink`;
  }
  return (
    <span className={containerCn}>
      {typedString}
      <span key={`${instanceId}-cur`} className={cursorCn}>
        {animationEnded && hideCursor ? '' : cursorCharacter}
      </span>
    </span>
  );
};

TypeMe.defaultProps = {
  onAnimationEnd: () => {},
  startAnimation: true,
  cursorCharacter: '|',
  cursorBlinkSpeed: 800, // ms
  backspaceDelay: 500, // ms
  typingSpeed: 500, // WPM
  deleteSpeed: 800, // WPM
  hideCursor: false,
  className: '',
  loop: false,
  strings: []
};

export default TypeMe;
