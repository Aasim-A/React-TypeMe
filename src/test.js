import React from 'react';
import { render, cleanup } from 'react-testing-library';
import TypeMe from './index';
import Text from './Text';

let div;

afterEach(() => {
  document.body.removeChild(div);
  div = null;
  cleanup();
});

/*
<Text />
  - change cursor
  - keep cursor after animation
  - hide cursor after animation
*/

describe('<Text /> component', () => {
  // render text passed as children
  it('<Text/> renders text passed as children', done => {
    new Promise(resolve => {
      const { container } = render(
        <Text
          onAnimationEnd={() => {
            resolve();
          }}
        >
          Hello!
        </Text>
      );
      div = container;
    }).then(() => {
      done();
      expect(div).toMatchSnapshot();
    });
  });

  // onAnimationEnd callback executed once successfully
  it(`<Text/> onAnimationEnd callback executed once successfully`, done => {
    let cb;
    new Promise(resolve => {
      cb = jest.fn(() => {
        resolve();
      });
      const { container } = render(<Text onAnimationEnd={cb}>Hello!</Text>);
      div = container;
    }).then(() => {
      done();
      expect(cb).toHaveBeenCalledTimes(1);
      expect(div).toMatchSnapshot();
      cb = null;
    });
  });

  // delete characters
  it('<Text/> renders text with deleted characters', done => {
    new Promise(resolve => {
      const { container } = render(
        <Text
          deleteCharacters={7}
          onAnimationEnd={() => {
            resolve();
          }}
        >
          Hello world!
        </Text>
      );
      div = container;
    }).then(() => {
      done();
      expect(div).toMatchSnapshot();
    });
  });

  // add lineBreak after render
  it('<Text/> renders with a line break character', done => {
    new Promise(resolve => {
      const { container } = render(
        <Text
          lineBreak={true}
          onAnimationEnd={() => {
            resolve();
          }}
        >
          Line break?
        </Text>
      );
      div = container;
    }).then(() => {
      done();
      expect(div).toMatchSnapshot();
    });
  });

  // trigger startAnimation manually
  it('<Text/> with manually triggered start animation', done => {
    new Promise(resolve => {
      const { container, rerender } = render(
        <Text startAnimation={false}>Nope</Text>
      );
      setTimeout(() => {
        rerender(
          <Text
            startAnimation={true}
            onAnimationEnd={() => {
              resolve();
            }}
          >
            Yes
          </Text>
        );  
      }, 500);
      div = container;
    }).then(() => {
      done();
      expect(div).toMatchSnapshot();
    });
  });

  // change default cursor
  it('<Text/> renders with new cursor character', done => {
    new Promise(resolve => {
      const { container } = render(
        <Text
          cursorCharacter="_"
          onAnimationEnd={() => {
            resolve();
          }}
        >
          New cursor?
        </Text>
      );
      div = container;
    }).then(() => {
      done();
      expect(div).toMatchSnapshot();
    });
  })
});

/* 
<TypeMe />
  - renders without crashing
  - renders multiple instances 
  - renders text passed as children
  - renders text passed as props
  - animationEnd callback executed once successfully
*/

describe('<TypeMe /> component', () => {
  // renders without crashing
  it('<TypeMe/> renders without crashing', done => {
    const { container } = render(<TypeMe />);
    setTimeout(() => {
      div = container;
      expect(div).toMatchSnapshot();
      done();
    }, 1000);
  });
});
