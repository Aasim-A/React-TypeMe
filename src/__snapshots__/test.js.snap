// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> component <Text/> onAnimationEnd callback executed once successfully 1`] = `
<div>
  <span
    class="ta-anim"
  >
    <span
      class=""
    >
      Hello!
    </span>
    <span
      class="ta-cursor ta-blink"
    >
      |
    </span>
  </span>
</div>
`;

exports[`<Text /> component <Text/> renders text passed as children 1`] = `
<div>
  <span
    class="ta-anim"
  >
    <span
      class=""
    >
      Hello!
    </span>
    <span
      class="ta-cursor ta-blink"
    >
      |
    </span>
  </span>
</div>
`;

exports[`<Text /> component <Text/> renders text with deleted characters 1`] = `
<div>
  <span
    class="ta-anim"
  >
    <span
      class=""
    >
      Hello
    </span>
    <span
      class="ta-cursor ta-blink"
    >
      |
    </span>
  </span>
</div>
`;

exports[`<Text /> component <Text/> renders with a line break character 1`] = `
<div>
  <span
    class="ta-anim"
  >
    <span
      class=""
    >
      Line break?
    </span>
    <span
      class="ta-cursor ta-blink"
    >
      |
    </span>
    <br />
  </span>
</div>
`;

exports[`<Text /> component <Text/> renders with new cursor character 1`] = `
<div>
  <span
    class="ta-anim"
  >
    <span
      class=""
    >
      New cursor?
    </span>
    <span
      class="ta-cursor ta-blink"
    >
      _
    </span>
  </span>
</div>
`;

exports[`<Text /> component <Text/> with manually triggered start animation 1`] = `
<div>
  <span
    class="ta-anim"
  >
    <span
      class=""
    >
      Yes
    </span>
    <span
      class="ta-cursor ta-blink"
    >
      |
    </span>
  </span>
</div>
`;

exports[`<TypeMe /> component <TypeMe/> renders without crashing 1`] = `<div />`;
